worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,237 @@
+				// Best we can do here is log an error.
+				w.Logger.Warningf("add metrics user failed, but could not clean up user %q: %v",
+					username, err)
+			}
+		}
+	}()
+
+	var model model
+	model, err = w.State.Model()
+	if err != nil {
+		return errors.Annotatef(err, "retrieving current model: %v", err)
+	}
+
+	_, err = model.AddUser(state.UserAccessSpec{
+		User:      userTag,
+		CreatedBy: names.NewUserTag(userCreator),
+		Access:    permission.ReadAccess,
+	})
+	if err != nil {
+		return errors.Annotatef(err, "adding user %q to model %q: %v", username, bootstrap.ControllerModelName, err)
+	}
+
+	return nil
+}
+
+func (w *Worker) handleRemoveMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodDelete {
+		resp.Header().Add("Allow", http.MethodDelete)
+		w.writeResponse(resp, http.StatusMethodNotAllowed, errorResp("method must be %q, got %q", http.MethodDelete, req.Method))
+		return
+	}
+
+	username := strings.TrimPrefix(req.URL.Path, "/metrics-users/")
+	err := w.removeMetricsUser(username)
+	if err != nil {
+		w.writeResponse(resp, asHTTPCode(err), errorResp(err.Error()))
+		return
+	}
+	w.writeResponse(resp, http.StatusOK, struct{}{})
+}
+
+func (w *Worker) removeMetricsUser(username string) error {
+	err := validateMetricsUsername(username)
+	if err != nil {
+		return err
+	}
+
+	userTag := names.NewUserTag(username)
+	err = w.State.RemoveUser(userTag)
+	if errors.Is(err, errors.NotFound) || errors.Is(err, errors.UserNotFound) {
+		// succeed as no-op
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateMetricsUsername(username string) error {
+	if username == "" {
+		return errors.BadRequestf("missing username")
+	}
+
+	if !names.IsValidUserName(username) {
+		return errors.NotValidf("username %q", username)
+	}
+
+	if !strings.HasPrefix(username, jujuMetricsUserPrefix) {
+		return errors.BadRequestf("username %q should have prefix %q", username, jujuMetricsUserPrefix)
+	}
+
+	return nil
+}
+
+func (w *Worker) writeResponse(resp http.ResponseWriter, statusCode int, body any) {
+	w.Logger.Debugf("operation finished with HTTP status %v", statusCode)
+	resp.WriteHeader(statusCode)
+
+	message, err := json.Marshal(body)
+	if err != nil {
+		w.Logger.Errorf("error marshalling response body to JSON: %v", err)
+		w.Logger.Errorf("response body was %#v", body)
+		// Just write an empty response
+		message = []byte("{}")
+	}
+
+	w.Logger.Tracef("returning response %q", message)
+	_, err = resp.Write(message)
+	if err != nil {
+		w.Logger.Warningf("error writing HTTP response: %v", err)
+	}
+}
+
+// errorResp returns a struct that marshals to
+//
+//	{"error": <provided error message>}
+func errorResp(format string, args ...any) any {
+	return struct {
+		Error string `json:"error"`
+	}{
+		Error: fmt.Sprintf(format, args...),
+	}
+}
+
+// asHTTPCode translates an error from the `juju/errors` package into an HTTP
+// status code.
+func asHTTPCode(err error) int {
+	switch {
```

Comment:
We should really only be returning the errors we explicitly handle, everything else should be 500s. As in, if you expect the business logic to return a NotFound, then you can pass that over the wire. But if the business logic gave you an unexpected Unauthorized or BadRequest, that isn't as meaningful to the client.